allennlp/modules/token_embedders/openai_transformer_embedder.py

```diff
@@ -0,0 +1,110 @@
+
+        self._transformer = transformer
+        self._num_output_representations = num_output_representations
+        self._do_layer_norm = do_layer_norm
+
+        self._scalar_mixes: List[ScalarMix] = []
+
+        for k in range(num_output_representations):
+            scalar_mix = ScalarMix(transformer.num_output_layers, do_layer_norm=do_layer_norm)
+            self.add_module(f'scalar_mix_{k}', scalar_mix)
+            self._scalar_mixes.append(scalar_mix)
+
+    def get_output_dim(self):
+        """
+        The last dimension of the output, not the shape.
+        """
+        return self._transformer.embed.embedding_dim
+
+    def forward(self, inputs: torch.Tensor, offsets: torch.Tensor) -> torch.Tensor:
+        """
+        Parameters
+        ----------
+        inputs: ``torch.Tensor``, required
+            A ``(batch_size, num_timesteps)`` tensor representing the byte-pair encodings
+            for the current batch.
+        offsets: ``torch.Tensor``, required
+            A ``(batch_size, max_sequence_length)`` tensor representing the word offsets
+            for the current batch.
+
+        Returns
+        -------
+        ``[torch.Tensor]``
+            An embedding representation of the input sequence
+            having shape ``(batch_size, sequence_length, embedding_dim)``
+        """
+        # pylint: disable=arguments-differ
+        batch_size, num_timesteps = inputs.size()
+
+        # the transformer "vocab" consists of the actual vocab and the
+        # positional encodings. Here we want the count of just the former.
+        vocab_size = self._transformer.vocab_size - self._transformer.n_ctx
+
+        # Combine the inputs with positional encodings
+        batch_tensor = torch.zeros((batch_size, num_timesteps, 2), dtype=torch.long)
+        batch_tensor[:, :, 0] = inputs
+        batch_tensor[:, :, 1] = torch.arange(vocab_size, vocab_size + num_timesteps)
+
+        byte_pairs_mask = inputs != 0
+
+        # Embeddings is num_output_layers x (batch_size, num_timesteps, embedding_dim)
+        layer_activations = self._transformer(batch_tensor)
+
+        # Output of each scalar_mix is (batch_size, num_timesteps, embedding_dim)
+        mixes = [scalar_mix(layer_activations, byte_pairs_mask) for scalar_mix in self._scalar_mixes]
+
+        # These embeddings are one per byte-pair, but we want one per original _word_.
+        # Fortunately, we have the ``offsets`` which indicate the last byte-pair for
+        # each original word. Here we just choose that last byte-pair, although you
+        # could imagine doing something more sophisticated.
+        _, max_sequence_length = offsets.size()
+        _, _, embedding_dim = mixes[0].size()
+
+        # TODO(joelgrus): vectorize?
+        embeddings = [torch.zeros(batch_size, max_sequence_length, embedding_dim) for _ in mixes]
+        for i in range(batch_size):
```

Comment:
You can vectorise this by doing:

```
range_vector = get_range_vector(0, batch_size).unsqueeze(1)

last_byte_pair_embeddings = embeddings[range_vector, offsets]

```
You can see an example of this [here](https://github.com/allenai/allennlp/blob/master/allennlp/models/biaffine_dependency_parser.py#L501).

This will save a lot of memory, I think.

Reply:
awesome, thanks